### DISCUSSION TITLE
Datasource with Kanban/Board/Table-View

### DISCUSSION THREAD

**Discussion by MDXDave** [outline/outline#3440] [Ideas]
It would be really helpful if Outline had the possibility of defining data sources and display them dynamically as a Kanban board or in a table, like notion does it.

I am not sure if something like this is on the roadmap, but i think many users would use this feature.

![image](https://user-images.githubusercontent.com/6099976/164753041-0452a2de-47e1-46f0-83e1-e92784bad420.png)
![image](https://user-images.githubusercontent.com/6099976/164752970-a5429d46-268b-462a-80ac-8758c2738d32.png)

**Comment by ecker00** [comment]
Yes please! Probably my favorite Notion feature which is hard to find an alternative to.

**Comment by Gwemox** [comment]
Very good idea !

**Comment by inf0matics** [comment]
vote++

**Comment by corysus** [comment]
I try many alternative but always return to Outline, only missing part is this Kanban/Board/Table view. 
Foscalboard have something like Notion kanban but is not good enough because writing description is limited and not good organised like in Outline, so +1 vote for this feature.

**Comment by julianbrennum** [comment]
Yes! This is really the only feature I feel is missing from outline.

**Comment by dsmrs** [comment]
It's a nice feature, and can be linked to the "tag" functionnality (update tag when moving from a column, ex: *doing*>*done*) see: https://github.com/outline/outline/discussions/4671

**Comment by almereyda** [comment]
Eventually first-class integrations with Grist, Baserow, Budibase, NocoDB, Saltcorn, Seatable or the likes, loosely coupled via action and trigger nodes in n8n, could temporarily help to work around this perceived limitation? An integration with another focussed product in the desired area (low/no code databases) instead may also help to avoid feature creep/bloat in Outline.

But it is to second that the Notion outliner-block-editor–as if one would mix Outline with LogSeq–appears to set a new standard for contextual content editing.

**Comment by SonnTakuu** [comment]
Please, I really need that to review my notes.

**Comment by s4l4x** [comment]
+1 native database/kanban view

**Comment by sergi0g** [comment]
This feature would be extremely helpful. This is also the most upvoted discussion here. Please consider adding support for this, otherwise I'll just move to AppFlowy which is planning to add web support in a few months. Appflowy is a better alternative to Notion.

**Comment by NicolasRitouet** [comment]
having this feature would definitively helps the migration from notion.
@tommoor any plan (short or long term)? knowing if that's planned would def help

**Reply by tommoor** [reply]
No plans to build this I’m afraid

**Reply by corysus** [reply]
In that case probably most of people who need this part will switch to https://affine.pro. This project is in heavy development and current have database for table view, also kanban board comming probably soon.

**Reply by ovizii** [reply]
> In that case probably most of people who need this part will switch to https://affine.pro. This project is in heavy development and current have database for table view, also kanban board comming probably soon.

affine is nice and all but afaik its local only and not a web app

**Reply by tommoor** [reply]
Absolutely – folks should use whatever software makes sense for their usecase, Outline isn't trying to be everything to everyone and I'm certainly not trying to build anything that would lean into the task management space

**Reply by corysus** [reply]
I'm really sorry that you don't have this in your roadmap, it would be really good if we had kanban and in general some kind of database in Outline, but what's there is, otherwise I use Outline every day and it's a really good app.

**Reply by herzkerl** [reply]
@tommoor While we're not in need of a "task management" app per se, it would be great to have a simple overview of all unchecked tasks (i.e. 'task list' items) of all documents the user has access to; preferably sorted/grouped by collection, then document. Please give it some thought :) I'd be willing to (partly) sponsor that feature, too!

**Reply by ArcticXWolf** [reply]
@tommoor Just as an additional thought: This feature would not restrict itself only to the "Task management" usecase. For example I plan and organize my D&D-Campaigns with Notion and would love to switch that to Outline. But I have a huge list of articles/pages for each location/character/item and need an overview "view". Thats why notions Table-View is awesome, because you can define tags on each arcticle (think like form fields) and see all of them in one place.

For an example, see here: https://slyflourish.notion.site/52bbcbf26e1e43ee80b8d69e92cb35e4?v=d91eab449fd140c9bbda26514a224f7a
![image](https://github.com/outline/outline/assets/5812215/8acb8a9a-8cbe-4aac-a1df-63e11cd885ff)

Just so we dont dismiss this feature because we do not want to be a "project/task management" software, this feature is a great way to organize notes of ANY kind. (Think of authors doing world building, think of cooking recipes, etc)

**Reply by ArcticXWolf** [reply]
So maybe the feature request should be less about tasks and kanban view, but more about:

1. A way to group similar pages (-> notion database)
2. A way to add machine-readable information to pages (-> notion form fields)
3. Multiple views to show and visualize the pages and info from 1. and 2. (-> notion table view, gallery view, maybe later kanban view)

**Reply by ciekawy** [reply]
Well IMO simple data tabes is a must have for this kind of software. Eg to include sth like this, with persistency and reference in the page - is it any difficult in react?

https://mui.com/material-ui/react-table/

https://github.com/jbetancur/react-data-table-component

**Reply by problem** [reply]
To re-state the summary from @ArcticXWolf , I think the essential points are:
1. Tags on Docs
2. The ability to query against Tags
3. Display results from these queries in a table

Just starting there would be great. All the different views, etc in Notion are what make it powerful, but that's also where Notion starts to get messy. In their purest form, I can get behind these requests and we would definitely have a use for them as our use of Outline grows at our org.

In general I think keeping the project management aspects of Notion out of Outline are essential to maintaining its essence as a bloat-free wiki that's very focused on only what's needed for a doc/wiki use-case.

**Reply by almereyda** [reply]
Just to add my two cents, Outline only wins by not trying to be a "project management solution", but being a good team wiki.

That doesn't mean #765 wouldn't be useful.

Maybe this kind of tabular (re)view of document lists filtered by tag could be integrated into #2273?

**Reply by wforte88** [reply]
I agree with @almereyda that Outline's focus does not seem to be on becoming a task manager, but the filtered database view goes far beyond a task manager. It allows for the creation of taxonomies for articles, which would elevate Outline to another level of CMS. It would be possible to manage more complex knowledge bases, in which there is not just one level of collection taxonomy.

**Comment by fvsion** [comment]
+1 for this feature

This is the one feature that would make me stay on outline and not even look elsewhere. Outline has been incredible, but this is a key missing feature that should be integrated, esp. with competitors (which are a bit more individual wiki focused whereas outline is a team wiki) like Affine and Appflowy already having this feature. 

Although... this does move it a bit outside the function of a wiki...but we already have "tasks" and outline is already a bit more than just a wiki. Just depends on that target market of the outline team.

I would argue that this doesn't really make notion a project management solution as Plane or Asana are much better for that.

**Comment by almereyda** [comment]
If Outline were ever to support such a feature, it would enter the competitive market around lo/no code environments. We find many degrees of complexity associated with that, sometimes just as a database shim, other times more as an (internal) app development platform.

<details><summary>These range from simple and easy to use interfaces</summary>
<p>

* [Mattermost Platform Overview | All Communications On One Platform](https://mattermost.com/platform-overview/)
  * [mattermost/focalboard: Focalboard is an open source, self-hosted alternative to Trello, Notion, and Asana.](https://github.com/mattermost/focalboard)
* [SeaTable: einfach wie Excel, mächtig wie eine Datenbank.](https://seatable.io/)
  * [seatable/seatable: SeaTable: easy like a spreadsheet, powerful like a database. Unlimited rows in a single base.](https://github.com/seatable/seatable/) (not FLOSS)
* [Tables - Apps - App Store - Nextcloud](https://apps.nextcloud.com/apps/tables)
  * [nextcloud/tables: 🍱 Nextcloud tables app](https://github.com/nextcloud/tables/)

</p>
</details> 

<details><summary>To more sophisticated low code environments that wrap a database</summary>
<p>

* [Baserow: Open source no-code database and Airtable alternative](https://baserow.io/)
  * [Baserow / baserow · GitLab](https://gitlab.com/baserow/baserow)
* [Supabase | The Open Source Firebase Alternative](https://supabase.com/)
  * [supabase/supabase: The open source Firebase alternative. Follow to stay updated about our public Beta.](https://github.com/supabase/supabase/)
* [NocoDB | Turns your SQL database into a Nocode platform. Free & Open Source.](https://nocodb.com/)
  * [nocodb/nocodb: 🔥 🔥 🔥 Open Source Airtable Alternative](https://github.com/nocodb/nocodb)

</p>
</details> 

<details><summary>To those which bring their own custom data store</summary>
<p>

* [Grist | The Evolution of Spreadsheets](https://www.getgrist.com/)
  * [gristlabs/grist-core: Grist is the evolution of spreadsheets.](https://github.com/gristlabs/grist-core/)
* [Budibase | Build internal tools in minutes, the easy way](https://budibase.com/)
  * [Budibase/budibase: Low code platform for creating internal tools, workflows, and admin panels in minutes. Supports PostgreSQL, MySQL, MSSQL, MongoDB, Rest API, Docker, K8s, and more 🚀. Budibase, the low code platform you'll enjoy using ⚡](https://github.com/budibase/budibase)

</p>
</details>

In the end we are looking for a generalisation of types of content that can be visualised in multiple manners. Some call this the "block protocol":

* [Block Protocol – an open standard for data-driven blocks](https://blockprotocol.org/)
* [HASH - a workspace for structured knowledge.](https://hash.ai/)
  * [hashintel/hash: 🚀 The best way to use and work with blocks](https://github.com/hashintel/hash) (the editor already works (partly), and it's fun to try out)

I fear this is a totally different scope for the application, but indeed also opens up many interesting possibilities.

The implementation in Focalboard / Mattermost Boards is distinctively low key and very simple to grasp. A bit like a Trello with custom fields, which can be displayed in tables, too. Maybe Outline could learn from its approach to provide just another view to data which is already there.

Ideally, as the request rate for this feature is relatively high, this could also be well worth a community effort to organise around a PoC implementation for showcasing its usefulness.

**Comment by zeekhuge** [comment]
A different approach would probably be to have a very tight integration with tools like Trello/Airtable.

The `/` command would have to allow 
- Logging into these tools through OAuth
- Creating new doc on these tools (which would be like creating a new Notion DB)
- Searching through the available pages (which individual user has access to) on these tools & embed them in the page.

The current integrations, however, are only front-end oriented i.e. they are basically just iframes & backend is not involved in any way - no OAuth, APIs, etc (& seems @tommoor wants to keep it that way ?).

But, the above suggested integration will require full backend support as well.

Any thoughts @tommoor ? Do you think Outline may try some tighter integration with these tools ?

**Reply by sergi0g** [reply]
Another idea would be an integration with Focalboard, which is like Notion databases and has API support.

**Comment by ArcticXWolf** [comment]
Hmm, maybe we should split this into multiple issues/features. I get the feeling that having "Kanban" in the post title lets everyone only focus on project management or on emulating nocode software.

But in my opinion we could start with much simpler features that make **organizing** Outline MUCH better:

1. Being able to group pages together and enforce their templates
2. Add metadata/organized data to such pages
3. (As a last addition) Create a nice overview of such a group

Two very simple usecases (**that do not target project management**):

1. I use Outline to store recipes. While I can use templates to make them look similar, it would be great to be able to add metadata to the pages like "how many servings" or "time to prep". For now I add a table at the beginning of each site, but this isnt very elegant (for example when I need to change the template). Also we could use these metadata to filter and search for recipes (by taste, style, number of servings or else).

2. I would love to use Outline for organizing DnD-Campaigns (to do worldbuilding). I have already described this in https://github.com/outline/outline/discussions/3440#discussioncomment-6305629


I dont think we have to create a full project management suite or nocode functionality for this :)

**Reply by zeekhuge** [reply]
The more I am thinking about it, the more it seems like datasource should be an internal feature & not just a tight integration with something else.

The ultimate reason why we want things to be on Outline is to get away from the hassle of maintaining, monitoring & moving multiple different tools. Prevent the context shifting. For companies - to explain, train & onboard employees on different tools.

I respect @tommoor's decision to have a well defined focused. But maybe we as a community can try to take this up.
What we can probably do is:
1. Create a separate micro-service which would be hosted in the same docker-compose as outline.
2. This service will use APIs (to make mods to the content) as much as it can, since we expect them to be pretty stable.
3. The things that cannot be done through APIs (in-app notifications, etc) can be extended by making this micro-service directly interact with the DB.
4. Fork Outline (or convince @tommoor) to just add 1 UI component (to be maintained by the community) to build the UI for this data-source.

Also, does the license allow us to do all this ?

**Reply by zeekhuge** [reply]
Also, this will probably need (seemingly planned - https://github.com/outline/outline/pull/5553) JSON based document instead of MD based.

**Reply by patzm** [reply]
I also agree that an external solution would defeat the core purpose of Outline (and probably why it is so successful). But indeed, focus is probably the most important aspect of the choice. I 2nd that support for @tommoor.

**Comment by Snuupy** [comment]
Just discovered this issue organically myself and it seems that this is the most requested feature for outline. https://github.com/outline/outline/discussions?discussions_q=is%3Aopen+sort%3Atop

**Comment by uxodb** [comment]
Same as above poster. Has there been any word on this after 2 years?

**Reply by keithschacht** [reply]
The developers chimed in already on the Sept 22nd, 2023 thread

**Comment by g-nardiello** [comment]
My use case will also fit to this discussion.

I have some data to be shared across different documents on the wiki, in a tabular view.

If a data table will be available, it could be added in each document with filtering on the right tags.

Also, it makes sense that is a native component because:
- I could interact with the table and go to the page linked to each record
- if I add a row on a data table in a specific document, all the documents will be updated with the new row (preventing inconsistencies)
- if I export the markdown document the table will be converted using the markdown syntax 

PS: as this is the most required feature, I think you could find help on contributors to get this feature implemented

**Comment by ghost** [comment]
Hey everyone! I’ve started a new discussion on adding **structured data (databases)** to Outline and referenced this thread among other related ideas.

The new post is quite visual and includes a current ecosystem overview, potential features, and mockups. I’d love your feedback:
⭐ #8476

**Comment by bvandevliet** [comment]
Yes pls!
Integrated native Kanban/Board snippet would be a killer feature! I currently use https://github.com/plankanban/planka next to Outline, because it's like a self-hosted alternative to Trello. But of course it's not connected to or integrating with Outline in any way, it's a completely separate app. I'm thinking native support to embed boards from a Planka instance would also be a very welcome feature, not as nice as built-in Kanban snippet, but still very nice, right? And maybe easier to implement? :)

**Comment by tsutsurocket** [comment]
Hi all,

I use Outline for my notes. As a side project I built a small project board on top of it. Sharing it here in case it is useful or interesting to anyone.

The idea: a board with Product > Epic > Story > Task/Issue, where every item is just an Outline document. No in between database. The board is a view over a collection.

How it works:
- The board loads from one `collections.documents` call. No per-card requests.
- The metadata sits in the document title, like `Fix login redirect loop [TASK][P2][In progress][@sam]`. So one call has everything needed to draw the board.
- The body is a normal Outline doc with history, search, comments and permissions. Click a card and it opens the real document.
- Changing status, priority, assignee or name is just a title update (`documents.update`). Moving a card is `documents.move`.
- It uses the public REST API. No fork needed.

A `[SKIP]` tag in the title hides scratch docs from the board.

What is next: I am adding AI agents that pickup jobs from the board. I assign a task to the AI agent, it reads the doc, writes the code, opens a PR, and posts back on the document.

<img width="1911" height="824" alt="image" src="https://github.com/user-attachments/assets/918067fe-4f57-4045-b2ef-56c89b56aacf" />

A couple of questions:
- Has anyone built a board or app view straight on the documents API? Keen to hear how you approached it.
- Anything obvious I am missing or should watch out for?

Cheers

**Reply by bvandevliet** [reply]
This looks cool! One idea that comes to mind is that you might want to use front matter for metadata instead (keeping the title clean and lean). And is this something you consider to open source? :)

**Reply by tsutsurocket** [reply]
I am sticking to self hosted & community code, hence no access to meta data currently

**Comment by T-Herrmann-WI** [comment]
Yes Kanban in Outline would be very nice.

**Comment by Plinsboorg** [comment]
Hi all. I've implemented a draft version of the document databases for Outline.
PR: https://github.com/outline/outline/pull/13210
Try it: https://outline-db-test.plinsburg.tech/
the magic email sign in will let you in 
<img width="1873" height="690" alt="image" src="https://github.com/user-attachments/assets/f186ab4d-9e5e-4301-8e5a-8475a5a4f92e" />

See this discussion for details
https://github.com/outline/outline/discussions/8476#discussioncomment-17851977

I'd love to hear from maintainers if this fully vibe coded solution has a chance to be accepted.
